### PR TITLE
Migrating data to newUserFunctionId tables as part of Consumption plan support

### PR DESCRIPTION
**Pull request by MaddyDev**
**New UserFunctionId**:
We have used the [WEBSITE_SITE_NAME](https://learn.microsoft.com/en-us/azure/app-service/reference-app-settings?tabs=kudu%2Cdotnet#app-environment) instead of the host id for creating the new **_userFunctionId_**

**Migration** steps followed:
1. While creating the lease table entry in the global state table, check if there a value with the oldUserFunctionId and use that value while inserting a new entry with the new userFunctionId to the global state table. And delete the oldUserFunctionId  entry.
2. Create new leases table with new hash and if old leases table exists, copy over all the data from it and delete the table.

**Test**: To Be Done

### CODE REVIEW

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 1]
```diff
@@ -389,16 +392,30 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
                 }
             }
 
-            string insertRowGlobalStateTableQuery = $@"
+            bool migrateOldUserFunctionIdData = await this.OldUserFunctionIdRecordExists(connection, transaction, userTableId, cancellationToken);
```
Comment (Charles-Gagnon) [inline src/TriggerBinding/SqlTriggerListener.cs:395]: Do we need to do this check? Couldn't you just make each of the below queries do the check regardless and migrate as needed? Would save a bit of execution time and duplication of the query.

Comment (MaddyDev) [reply src/TriggerBinding/SqlTriggerListener.cs:395]: The reason I added the method is to add a telemetry event to track migration which I totally missed to add here. Will look into if there is a way to not have this check and still track migration.

Comment (Charles-Gagnon) [reply src/TriggerBinding/SqlTriggerListener.cs:395]: Ok, although given how little we look at this stuff currently I'd question whether we actually care about tracking this. Is that something you see being valuable?

Anyways it's fine if you want to keep this, not a huge deal (and we do other similar things elsewhere anyways).

Comment (MaddyDev) [reply src/TriggerBinding/SqlTriggerListener.cs:395]: I don't really think that tracking is required here but definitely useful when deciding on deprecation path, we can look if there are any still using the migration process and take a call. Plus Chris asked me to add a telemetry point :)

Comment (Charles-Gagnon) [reply src/TriggerBinding/SqlTriggerListener.cs:395]: We could just look at version numbers for that, right? It's not like people can opt out of this so as long as everyone is on the version with this change or higher we should be able to assume they've migrated

File: src/TriggerBinding/SqlTriggerBinding.cs [hunk 2]
```diff
@@ -95,14 +96,37 @@ public ParameterDescriptor ToParameterDescriptor()
         /// <summary>
         /// Returns an ID that uniquely identifies the user function.
         ///
+        /// We call the WEBSITE_SITE_NAME from the configuration and use that to create the hash of the
+        /// user function id. Appending another hash of class+method in here ensures that if there
+        /// are multiple user functions within the same process and tracking the same SQL table, then each one of them
+        /// gets a separate view of the table changes.
+        /// </summary>
+        private string GetUserFunctionIdAsync()
```
Comment (alrod) [inline src/TriggerBinding/SqlTriggerBinding.cs:104]: Remove Async from the name

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 3]
```diff
@@ -426,18 +443,41 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
         {
             string primaryKeysWithTypes = string.Join(", ", primaryKeyColumns.Select(col => $"{col.name.AsBracketQuotedString()} {col.type}"));
             string primaryKeys = string.Join(", ", primaryKeyColumns.Select(col => col.name.AsBracketQuotedString()));
+            string OldLeasesTableName = leasesTableName.Contains(this._userFunctionId) ? leasesTableName.Replace(this._userFunctionId, this._oldUserFunctionId) : string.Empty;
```
Comment (alrod) [inline src/TriggerBinding/SqlTriggerListener.cs:446]: oldLeasesTableName

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 4]
```diff
@@ -391,19 +394,33 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
 
             string insertRowGlobalStateTableQuery = $@"
                 {AppLockStatements}
-
+                
                 IF NOT EXISTS (
                     SELECT * FROM {GlobalStateTableName}
                     WHERE UserFunctionID = '{this._userFunctionId}' AND UserTableID = {userTableId}
                 )
+                BEGIN
```
Comment (alrod) [inline src/TriggerBinding/SqlTriggerListener.cs:402]: Could you please add a comment explaining why do we need _oldUserFunctionId and whats is GlobalStateTableName

Comment (MaddyDev) [reply src/TriggerBinding/SqlTriggerListener.cs:402]: Yes, this for back compatibility. We have the version at which the function is in the GlobalState table, this is to copy that version from the `oldUserFunctionId` to the new `userFunctionId` and continue the function.

File: src/SqlBindingConstants.cs [hunk 5]
```diff
@@ -7,6 +7,8 @@ internal static class SqlBindingConstants
     {
         public const string ISO_8061_DATETIME_FORMAT = "yyyy'-'MM'-'dd'T'HH':'mm':'ss'.'fffZ";
 
+        public const string WebsiteName = "WEBSITE_SITE_NAME";
```
Comment (chlafreniere) [inline src/SqlBindingConstants.cs:10]: nit: all uppercase variable name for const

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 6]
```diff
@@ -426,18 +443,41 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
...
+                        {LeasesTableChangeVersionColumnName} bigint NOT NULL,
+                        {LeasesTableAttemptCountColumnName} int NOT NULL,
+                        {LeasesTableLeaseExpirationTimeColumnName} datetime2,
+                        PRIMARY KEY ({primaryKeys})
+                    );
+            " : $@"
                 {AppLockStatements}
 
                 IF OBJECT_ID(N'{leasesTableName}', 'U') IS NULL
+                Begin
```
Comment (chlafreniere) [inline src/TriggerBinding/SqlTriggerListener.cs:463]: Let's remain consistent and use all uppercase for TSQL

File: src/SqlBindingUtilities.cs [hunk 7]
```diff
@@ -54,6 +54,15 @@ public static string GetConnectionString(string connectionStringSetting, IConfig
             return connectionString;
         }
 
+        public static string GetWebSiteName(IConfiguration configuration)
+        {
+            if (configuration == null)
```
Comment (chlafreniere) [inline src/SqlBindingUtilities.cs:59]: What happens for local development? Is website site name set?

File: src/Telemetry/Telemetry.cs [hunk 8]
```diff
@@ -348,7 +348,8 @@ public enum TelemetryEventName
         TableInfoCacheMiss,
         TriggerFunction,
         TriggerMonitorStart,
-        Upsert
+        Upsert,
+        InsertGlobalStateTableRow,
```
Comment (chlafreniere) [inline src/Telemetry/Telemetry.cs:352]: nit: unnecessary extra `,`

File: src/TriggerBinding/SqlTriggerBinding.cs [hunk 9]
```diff
@@ -78,8 +78,9 @@ public async Task<IListener> CreateListenerAsync(ListenerFactoryContext context)
         {
             _ = context ?? throw new ArgumentNullException(nameof(context), "Missing listener context");
 
-            string userFunctionId = await this.GetUserFunctionIdAsync();
-            return new SqlTriggerListener<T>(this._connectionString, this._tableName, this._leasesTableName, userFunctionId, context.Executor, this._sqlOptions, this._logger, this._configuration);
+            string userFunctionId = this.GetUserFunctionId();
```
Comment (chlafreniere) [inline src/TriggerBinding/SqlTriggerBinding.cs:81]: Even after a migration is done once, are we still calculating these old + new values? Is that really needed?

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 10]
```diff
@@ -391,19 +394,33 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
...
             string insertRowGlobalStateTableQuery = $@"
                 {AppLockStatements}
-
+                -- For back compatibility copy the lastSyncVersion from _oldUserFunctionId if it exists.
                 IF NOT EXISTS (
                     SELECT * FROM {GlobalStateTableName}
                     WHERE UserFunctionID = '{this._userFunctionId}' AND UserTableID = {userTableId}
                 )
+                BEGIN
+                    -- Migrate LastSyncVersion from oldUserFunctionId if it exists and delete the record
```
Comment (chlafreniere) [inline src/TriggerBinding/SqlTriggerListener.cs:403]: Please use all caps for T-SQL

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 11]
```diff
@@ -391,19 +394,33 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
...
+                    Declare @lastSyncVersion bigint;
+                    Select @lastSyncVersion = LastSyncVersion from az_func.GlobalState where UserFunctionID = '{this._oldUserFunctionId}' AND UserTableID = {userTableId}
```
Comment (chlafreniere) [inline src/TriggerBinding/SqlTriggerListener.cs:405]: Plz check for sql injection possibilities

**Comment by chlafreniere** [comment]
@Charles-Gagnon can you take another look?

File: src/TriggerBinding/SqlTriggerListener.cs [hunk 12]
```diff
@@ -391,19 +394,33 @@ private async Task<long> InsertGlobalStateTableRowAsync(SqlConnection connection
...
+                    IF @lastSyncVersion is NULL
+                        SET @lastSyncVersion = {(long)minValidVersion};
+                    ELSE
+                        Delete from az_func.GlobalState where UserFunctionID = '{this._oldUserFunctionId}' AND UserTableID = {userTableId}
+                    
                     INSERT INTO {GlobalStateTableName}
                     VALUES ('{this._userFunctionId}', {userTableId}, {(long)minValidVersion}, GETUTCDATE());
```
Comment (Charles-Gagnon) [inline src/TriggerBinding/SqlTriggerListener.cs:412]: Shouldn't this be using @lastSyncVersion? That'll be the old value is we're migrating over.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
There's one potential bug to look at, and then please make sure to respond to Chris' comments as well. Otherwise looks good though!

File: src/SqlBindingUtilities.cs (hunk 7 above)
Comment (MaddyDev) [reply src/SqlBindingUtilities.cs:59]: WEBSITE_SITE_NAME value is null for local development.

File: src/TriggerBinding/SqlTriggerBinding.cs (hunk 9 above)
Comment (MaddyDev) [reply src/TriggerBinding/SqlTriggerBinding.cs:81]: Yes, we decided to have the queries such that if old values are present, migrate them and clean up else continue with new values.

**Review by Charles-Gagnon — APPROVED** [review]
I would definitely recommend updating the [README](https://github.com/Azure/azure-functions-sql-extension/blob/main/docs/TriggerBinding.md#az_funcleasestablename)/MS docs to make it clear how the app name relates to the table and the consequences of changing that. 

One thing I'm not sure we tested before is whether changing the app name previously would cause the host ID to change (and thus "lose" the history). I believe that would have been the case even previously, but now it's much more explicit so it'd be good to add that information in.

**Review by chlafreniere — APPROVED** [review]

**Comment by MaddyDev** [comment]
> I would definitely recommend updating the [README](https://github.com/Azure/azure-functions-sql-extension/blob/main/docs/TriggerBinding.md#az_funcleasestablename)/MS docs to make it clear how the app name relates to the table and the consequences of changing that.
> 
> One thing I'm not sure we tested before is whether changing the app name previously would cause the host ID to change (and thus "lose" the history). I believe that would have been the case even previously, but now it's much more explicit so it'd be good to add that information in.

Will be adding the docs along with an end to end test in a separate PR. Thanks for calling this out.

**Pull request merged**